### DISCUSSION TITLE
Why doesn't my socket.io service work on vercel?

### DISCUSSION THREAD

**Discussion by williangrios** [vercel/community#5016] [Feature Question]
### Question

Why doesn't my socket.io service work on vercel?
On Heroku it is working correctly (
with the frontend hosted on vercel), but I want to put it also on vercel. Because it does not work? Can anyone give me any tips?

This is a backend project separate from the frontend.

----- CODE
const { addUser, removeUserBySocketId, getUser, users} = require('./users.js')
const express = require("express");
const socketio = require("socket.io");
const http = require("http");

const PORT = process.env.PORT || 8900

const app = express();
const server = http.createServer(app);
const router = require('./router');
app.use(router);

const io = socketio(server,   {cors: {
  origin: "*", // Substitua pelo domínio onde seu código está sendo executado 
  methods: ["GET", "POST", "OPTIONS"]
}})

io.on("connection", (socket) => {
  socket.on("addUser", (userId) => {
    addUser(userId, socket.id);
    io.emit("getUsers", users);
  });

  socket.on(
    "sendMessage",
    ({ conversationId, senderId, receiverId, text, attachedImage, personalizedProposalId }) => {
      const user = getUser(receiverId);
      if (!user ) {
        return
      }
      io.to(user?.socketId).emit("getMessage", {
        conversationId,
        senderId,
        text,
        attachedImage,
        personalizedProposalId
      });
    }
  );

  socket.on("disconnect", () => {
    removeUserBySocketId(socket.id)
    io.emit("getUsers", users);
  });
});

server.listen(PORT, () => {
  console.log("SERVER RUNNING on--->", server.address(), ' PORT:', PORT);
});

Thanks, friends

**Comment by amyegan** [comment]
Hey Willian! Edge Functions and Serverless Functions do not support acting as a WebSocket server. Because these functions are stateless and have a maximum execution duration, it is not possible to maintain a WebSocket connection.

Third-party solutions are available to enable realtime communication: https://vercel.com/guides/publish-and-subscribe-to-realtime-data-on-vercel#providers

https://vercel.com/guides/do-vercel-serverless-functions-support-websocket-connections
